### PR TITLE
Add SmartRTB bidder

### PR DESCRIPTION
**Pull request by AndriyPavlyuk**
Add SmartRTB bidder to keep up with GO version.

### CODE REVIEW

File: src/main/resources/static/bidder-params/smartrtb.json [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+      "type": "string",
+      "description": "Specific zone ID for this placement, belonging to app/site",
+      "minLength": 20
+    },
+    "forceBid": {
+      "type": "boolean",
+      "description": "Force bids with a test creative"
+    }
+  },
+  "required": [ "publisherId" ]
```
[the remark this answers is not in the document]
Comment (AndriyPavlyuk) [reply src/main/resources/static/bidder-params/smartrtb.json]: It seems to be better name "publisherId" than just pub_id ?

[the remark this answers is not in the document]
Comment (DGarbar) [reply src/main/resources/static/bidder-params/smartrtb.json]: This json schema is used for parameter validation.
In GO they have absolutely identical schema, so you can just copy form there. 
It is not our decision how this parameters will be named in a received request.

File: src/test/java/org/prebid/server/bidder/smartrtb/SmartrtbBidderTest.java [hunk 2]
```diff
@@ -129,112 +73,81 @@ public void makeHttpRequestsShouldSkipInvalidImpressionAndAddError() {
                         "SmartRTB only supports banner and video"));
         assertThat(result.getValue()).hasSize(1)
                 .extracting(httpRequest -> mapper.readValue(httpRequest.getBody(), BidRequest.class))
-                .flatExtracting(BidRequest::getImp)
-                .containsOnly(imp);
+                .flatExtracting(BidRequest::getImp);
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/bidder/smartrtb/SmartrtbBidderTest.java]: What are you testing in this case ?

File: src/test/java/org/prebid/server/bidder/smartrtb/SmartrtbBidderTest.java [hunk 3]
```diff
@@ -129,112 +73,81 @@ public void makeHttpRequestsShouldSkipInvalidImpressionAndAddError() {
...
     }
 
     @Test
-    public void makeHttpRequestShouldReturnErrorIfImpFormatIsNull() {
+    public void  makeHttpRequestsShouldReturnErrorIfImpExtCouldNotBeParsed() {
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/bidder/smartrtb/SmartrtbBidderTest.java]: Ctrl + Alt + L

File: src/test/java/org/prebid/server/bidder/smartrtb/SmartrtbBidderTest.java [hunk 4]
```diff
@@ -38,7 +41,7 @@
 
 public class SmartrtbBidderTest extends VertxTest {
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/bidder/smartrtb/SmartrtbBidderTest.java]: Add test for validate 'req.ext' change.

File: src/main/java/org/prebid/server/bidder/smartrtb/SmartrtbBidder.java [hunk 5]
```diff
@@ -119,73 +120,66 @@ private ExtImpSmartrtb parseImpExt(Imp imp) {
...
-            }
-        }
-        return headers;
-    }
-
     @Override
     public Result<List<BidderBid>> makeBids(HttpCall<BidRequest> httpCall, BidRequest bidRequest) {
         if (httpCall.getResponse().getStatusCode() == HttpResponseStatus.NO_CONTENT.code()) {
             return Result.of(Collections.emptyList(), Collections.emptyList());
+        } else if (httpCall.getResponse().getStatusCode() == HttpResponseStatus.BAD_REQUEST.code()) {
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/smartrtb/SmartrtbBidder.java]: `httpCall.getResponse().getStatusCode()` - make variable

File: src/main/java/org/prebid/server/bidder/smartrtb/SmartrtbBidder.java [hunk 6]
```diff
@@ -119,73 +120,66 @@ private ExtImpSmartrtb parseImpExt(Imp imp) {
...
+                final ObjectNode ext = bid.getExt();
+                if (ext == null) {
+                    return Result.emptyWithError(BidderError.badServerResponse(String.format(
+                            "Invalid bid extension from endpoint.")));
+                }
+                try {
+                    final SmartrtbResponseExt smartrtbResponseExt = Json.mapper.treeToValue(ext,
+                            SmartrtbResponseExt.class);
+
+                    BidType bidType = null;
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/smartrtb/SmartrtbBidder.java]: remove null

File: src/main/java/org/prebid/server/bidder/smartrtb/SmartrtbBidder.java [hunk 7]
```diff
@@ -119,73 +120,66 @@ private ExtImpSmartrtb parseImpExt(Imp imp) {
...
 
-    private static Result<List<BidderBid>> extractBids(BidRequest bidRequest, BidResponse bidResponse) {
-        if (bidResponse == null || bidResponse.getSeatbid() == null) {
-            return Result.of(Collections.emptyList(), Collections.emptyList());
+        final List<BidderBid> bidderBids = new ArrayList<>();
+        for (SeatBid seatBid : bidResponse.getSeatbid()) {
+            for (Bid bid : seatBid.getBid()) {
+                final ObjectNode ext = bid.getExt();
+                if (ext == null) {
+                    return Result.emptyWithError(BidderError.badServerResponse(String.format(
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/smartrtb/SmartrtbBidder.java]: String.format is redundant

File: src/main/java/org/prebid/server/bidder/smartrtb/SmartrtbBidder.java [hunk 8]
```diff
@@ -119,73 +120,66 @@ private ExtImpSmartrtb parseImpExt(Imp imp) {
...
+                            "Invalid bid extension from endpoint.")));
+                }
+                try {
+                    final SmartrtbResponseExt smartrtbResponseExt = Json.mapper.treeToValue(ext,
+                            SmartrtbResponseExt.class);
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/smartrtb/SmartrtbBidder.java]: You can extract this and ext==null to another validation method, to make it cleaner

File: src/main/java/org/prebid/server/bidder/smartrtb/SmartrtbBidder.java [hunk 9]
```diff
@@ -119,73 +120,66 @@ private ExtImpSmartrtb parseImpExt(Imp imp) {
...
+                            break;
+                        case CREATIVE_TYPE_VIDEO:
+                            bidType = BidType.video;
+                            break;
+                        default:
+                            return Result.emptyWithError(BidderError.badServerResponse(String.format(
+                                    "Unsupported creative type %s.", smartrtbResponseExt.getFormat())));
+                    }
+                    SmartrtbResponseExt.of(null);
+                    final BidderBid bidderBid = BidderBid.of(bid, bidType, DEFAULT_BID_CURRENCY);
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/smartrtb/SmartrtbBidder.java]: replace with
```
 final Bid updatedBid = bid.toBuilder().ext(null).build();
 final BidderBid bidderBid = BidderBid.of(updatedBid, bidType, DEFAULT_BID_CURRENCY);
   ```

File: src/test/java/org/prebid/server/bidder/smartrtb/SmartrtbBidderTest.java [hunk 10]
```diff
@@ -129,155 +124,69 @@ public void makeHttpRequestsShouldSetExpectedRequestUrlAndDefaultHeaders() {
...
+                mapper.writeValueAsString(givenBidResponse(bidBuilder -> bidBuilder.ext(ext))));
 
         // when
         final Result<List<BidderBid>> result = smartrtbBidder.makeBids(httpCall, null);
 
         // then
-        assertThat(result.getErrors()).isEmpty();
-        assertThat(result.getValue()).isEmpty();
+        assertThat(result.getErrors().get(0).getMessage()).startsWith("Unsupported creative type wrong type");
+        assertThat(result.getErrors().get(0).getType()).isEqualTo(BidderError.Type.bad_server_response);
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/bidder/smartrtb/SmartrtbBidderTest.java]: BidderError.badServerResponse("Unsupported creative type wrong type")

File: src/test/java/org/prebid/server/bidder/smartrtb/SmartrtbBidderTest.java [hunk 11]
```diff
@@ -129,155 +124,69 @@ public void makeHttpRequestsShouldSetExpectedRequestUrlAndDefaultHeaders() {
...
+                mapper.writeValueAsString(givenBidResponse(bidBuilder -> bidBuilder.impid("123"))));
 
         // when
         final Result<List<BidderBid>> result = smartrtbBidder.makeBids(httpCall, null);
 
         // then
-        assertThat(result.getErrors()).isEmpty();
+        assertThat(result.getErrors()).hasSize(1);
+        assertThat(result.getErrors().get(0).getMessage()).startsWith("Invalid bid extension from endpoint.");
+        assertThat(result.getErrors().get(0).getType()).isEqualTo(BidderError.Type.bad_server_response);
```
Comment (DGarbar) [inline src/test/java/org/prebid/server/bidder/smartrtb/SmartrtbBidderTest.java]: BidderError.badServerResponse("")

File: src/main/java/org/prebid/server/bidder/smartrtb/SmartrtbBidder.java [hunk 12]
```diff
@@ -1,9 +1,23 @@
...
+import org.apache.commons.lang3.StringUtils;
+import org.prebid.server.bidder.Bidder;
+import org.prebid.server.bidder.model.BidderBid;
+import org.prebid.server.bidder.model.BidderError;
+import org.prebid.server.bidder.model.HttpCall;
+import org.prebid.server.bidder.model.HttpRequest;
+import org.prebid.server.bidder.model.Result;
+import org.prebid.server.bidder.smartrtb.model.SmartrtbResponseExt;
+import org.prebid.server.exception.PreBidException;
+import com.fasterxml.jackson.core.JsonProcessingException;
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/smartrtb/SmartrtbBidder.java]: Please update your IDE code style. 
```
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.iab.openrtb.request.BidRequest;
import com.iab.openrtb.request.Imp;
import com.iab.openrtb.response.Bid;
import com.iab.openrtb.response.BidResponse;
import com.iab.openrtb.response.SeatBid;
import io.netty.handler.codec.http.HttpResponseStatus;
import io.vertx.core.MultiMap;
import io.vertx.core.http.HttpMethod;
import io.vertx.core.json.DecodeException;
import io.vertx.core.json.Json;
import org.apache.commons.lang3.StringUtils;
import org.prebid.server.bidder.Bidder;
import org.prebid.server.bidder.model.BidderBid;
import org.prebid.server.bidder.model.BidderError;
import org.prebid.server.bidder.model.HttpCall;
import org.prebid.server.bidder.model.HttpRequest;
import org.prebid.server.bidder.model.Result;
import org.prebid.server.bidder.smartrtb.model.SmartrtbResponseExt;
import org.prebid.server.exception.PreBidException;
import org.prebid.server.proto.openrtb.ext.ExtPrebid;
import org.prebid.server.proto.openrtb.ext.request.smartrtb.ExtImpSmartrtb;
import org.prebid.server.proto.openrtb.ext.request.smartrtb.ExtRequestSmartrtb;
import org.prebid.server.proto.openrtb.ext.response.BidType;
import org.prebid.server.util.HttpUtil;
```

File: src/main/java/org/prebid/server/bidder/smartrtb/SmartrtbBidder.java [hunk 13]
```diff
@@ -58,19 +61,17 @@ public SmartrtbBidder(String endpointUrl) {
...
                 final Imp validImp = validateImp(imp);
                 final ExtImpSmartrtb extImp = parseImpExt(imp);
 
-                if (StringUtils.isEmpty(pubId) && StringUtils.isNoneEmpty(extImp.getPubId())) {
+                if (StringUtils.isBlank(pubId) && StringUtils.isNoneEmpty(extImp.getPubId())) {
                     pubId = extImp.getPubId();
                 }
 
-                zoneId = extImp.getZoneId();
+                String zoneId = extImp.getZoneId();
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/smartrtb/SmartrtbBidder.java]: final

File: src/main/java/org/prebid/server/bidder/smartrtb/SmartrtbBidder.java [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+            };
+
+    private static final String DEFAULT_BID_CURRENCY = "USD";
+
+    private final String endpointUrl;
+
+    private static final String CREATIVE_TYPE_BANNER = "BANNER";
+
+    private static final String CREATIVE_TYPE_VIDEO = "VIDEO";
+
```
Comment (DGarbar) [inline src/main/java/org/prebid/server/bidder/smartrtb/SmartrtbBidder.java]: private static final String DEFAULT_BID_CURRENCY = "USD";
    private static final String CREATIVE_TYPE_BANNER = "BANNER";
    private static final String CREATIVE_TYPE_VIDEO = "VIDEO";
    private final String endpointUrl;

**Pull request merged**